### PR TITLE
[Feat]: Create Product API Endpoint

### PR DESCRIPTION
**Pull request by Neche-Stephen**
### How should this be manually tested?
---

Access the POST endpoint at /api/v1/products with a request body containing name, description, price, category
• Expected response are the required fields in JSON format. 
```json
{
    "status": "success",
    "message": "Product created successfully",
    "data": {
        "name": "sugar",
        "description": "sweet",
        "category": "sugarcane",
        "price": 100.0,
        "createdAt": "2024-07-29T08:40:33.368843",
        "updatedAt": "2024-07-29T08:40:33.370229",
        "id": "5ea3a9b0-5a17-4995-bed7-d29ebb3232ab"
    }
}
```
Access the POST endpoint at /api/v1/products with an incomplete request body
• Expected response is an appropriate error message and status code.
```json
{
  "status_code": 422,
  "status": "Bad request",
  "message": "Invalid input data",
  "errors": {
    "price": "Price must be a positive number"
  }
}
```

### Checklist of what you did
---

- [x] Implemented a functional POST endpoint at /api/v1/products endpoint.
- [x] Ensured endpoint successfully creates a new product and stores it in the database.
- [x] Appropriate HTTP status codes are returned (e.g., 201 Created for success, 400/422 Bad Request for invalid input).
- [x]  Ensured validation of input data to prevent malformed or incomplete product entries.

### Reference Issue
---

[FEAT]: Create Product API Endpoint ([#191](https://github.com/hngprojects/hng_boilerplate_java_web/issues/191))

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/product/controller/ProductController.java [hunk 1]
```diff
@@ -61,4 +62,44 @@ public ResponseEntity<ProductSearchDTO> searchProducts(
...
+        ProductStockDTO productStockDTO = new ProductStockDTO(
+                product.getId(),
+                product.getCurrentStock(),
+                product.getLastUpdated()
+        );
+        return new ResponseEntity<>(productStockDTO, HttpStatus.OK);
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/product/controller/ProductController.java:79]: The role based access here gives only the user with user role the authorization to create a product and an admin will not be able to create a product. Is this the expected behaviour?

File: src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java [hunk 2]
```diff
@@ -37,4 +39,24 @@ public ResponseEntity<?> handleException(NoHandlerFoundException ex){
...
+    @ExceptionHandler
+    public ResponseEntity<?> handleException(ProductNotFoundException ex){
+        productErrorDTO.setMessage(ex.getMessage());
+        productErrorDTO.setStatus_code(HttpStatus.NOT_FOUND.value());
+        productErrorDTO.setSuccess(false);
+        return new ResponseEntity<>(productErrorDTO, HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<?> handleValidationException(MethodArgumentNotValidException ex) {
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java:52]: An exception handler that handles methodargumentnotvalid already exists in the applicaition

File: src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
+import hng_java_boilerplate.user.repository.UserRepository;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
 import java.util.List;
+import java.util.UUID;
 
 @Service
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java:20]: remove this extra space

File: src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java (hunk 2 above)
Comment (Neche-Stephen) [reply src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java:52]: Please, I did not see that method in "ProductErrorHandler"

File: src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java (hunk 3 above)
Comment (Neche-Stephen) [reply src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java]: Removed.

File: src/main/java/hng_java_boilerplate/product/controller/ProductController.java [hunk 4]
```diff
@@ -61,4 +62,44 @@ public ResponseEntity<ProductSearchDTO> searchProducts(
...
+        ProductStockDTO productStockDTO = new ProductStockDTO(
+                product.getId(),
+                product.getCurrentStock(),
+                product.getLastUpdated()
+        );
+        return new ResponseEntity<>(productStockDTO, HttpStatus.OK);
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_USER') or hasRole('ROLE_ADMIN')")
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/product/controller/ProductController.java:79]: If any authenticated user should be able to create a product, is the preAuthorize needed?

Comment (lurldgbodex) [reply src/main/java/hng_java_boilerplate/product/controller/ProductController.java:79]: If only the admin should be able to create a product, remove the role_user, but if a user should be able to create product and is not limited to admin, remove the preauthorize, it's not needed

File: src/main/resources/db/migration/V13__alter_products_table_schema.sql [hunk 5]
```diff
@@ -0,0 +1,6 @@
+--Added needed columns
```
Comment (lurldgbodex) [inline src/main/resources/db/migration/V13__alter_products_table_schema.sql:1]: you also need to update the version

**Pull request closed without merging**